### DISCUSSION TITLE
How to use Extended User Model for Roles,

### DISCUSSION THREAD

**Discussion by ddoddsr** [spatie/laravel-permission#2667] [Q&A]
The User model is extended to a Client Model `class Client extends  User`
I want to use different policies for each.

Laravel 11
v6 permissions
standard migration, just id , No uuid

I am using this in both Model hoping that would be helpful, not sure if I need both:
`    protected function getDefaultGuardName(): string { return 'web'; }`
`    protected $guard_name = 'web';`

Client is unable to login with a Client Model with the entry for the /App/Models/Clients
but after **I manually edit the record** to ` /App/Models/User ` the client can login 

I can Assign the Role to the Client Model but it is not honored.

Do I need an entry in auth.php under providers?

**Comment by drbyte** [comment]
What database table are your Client models being stored in? Does it make a difference if you set `protected $table = 'clients';` in your Clients model? (and create the corresponding table)?

**Reply by ddoddsr** [reply]
I use the user table for the clients model.
`protected $table = 'users';`

It makes sense to separate it and make a clients table. 
Is that how most people handle this?

**Reply by drbyte** [reply]
I think what you're running into then is Laravel's default morph-map correlating the model with its table name.

You "can" override the morph mapping. One caveat I've run into with that is once I declared a custom morph map I had to manually declare "all" morph maps, not just my custom one. I never dug further to figure out how to keep "the rest" as dynamic.

**Comment by ddoddsr** [comment]
I'm trying the clients table
added the ` protected $table = 'clients'; `   to the Client Model
, but a client still cannot login.

Morph Map I have heard of but don't understand.
Have to do some reading.

**Comment by drbyte** [comment]
I wonder if it's as simple as setting the morph class for the model?
https://spatie.be/docs/laravel-permission/v6/advanced-usage/extending#content-child-user-models

**Reply by ddoddsr** [reply]
I will check it out, thanks!

**Comment by ddoddsr** [comment]
I tried seperate tables with the Client Model Extending the User Model.
and I have an entry in the auth.php  
`        'clients' => [
            'driver' => 'eloquent',
            'model' => App\Models\Client::class,
        ],`
Maybe I don't extend user but use the same traits & etc?

**Reply by ddoddsr** [reply]
Tried not extening user but use the same traits & etc.

Client still cannot login.

Permissions are made but not applied. How do I cause the user table to be used?
Is it in the Gate?

**Reply by drbyte** [reply]
You've mentioned making various changes to code to make it "work". Have you also been deleting previously-assigned role/permission assignments for your users? 
Because the morph class mapping is persisted in the database with each assignment.
So, if the lookup for permission access doesn't find a match in the db, access will not be granted.

**Reply by ddoddsr** [reply]
Yes I am rerunning the migration:fresh --seed for each change.
erified that the morph map was added as required..

This issue stems from wanting a UserPolicy and a ClientPolicy to assign diferent permissions to the User / Client Models for some roles. 
Maybe there is a way to achieve that extending the User Model

**Reply by drbyte** [reply]
In this package's test suite we have multiple User models, extending others for various reasons, and all the tests pass. Most of the tests run multiple times in various configurations for the sake of testing various custom model scenarios.

So I'm surprised at there being a problem.

**Reply by drbyte** [reply]
Also ... just to be clear: though a Client can extend a User, no Client "is" the same as a User: the models are distinctly separate, and have unique role/permission profiles: there's no inheritance between Users and Clients when it comes to roles/permissions .... er, well, I guess unless you were to stick both models into the same db table, in which case there could be unpredicted outcomes. 
Retrieval and caching are model-specific, all dependent on Laravel's Eloquent relations.

**Reply by ddoddsr** [reply]
I'm sure the problem is due to my lack of understanding.
I'll look at the test models
It sounds like I need to 

- Add new class Client extends User. Looking at the tests, Admin Model extending might be the example you are talking about. 
- add Policies for both.
- add a client migration table
- add enforceMorphMap to boot in AppServiceProvider
- create a guard 'client'   (that I will use to add permissions) ?  I'm OK with same guard at this point
- add a provider for client guard that points to the Client Model

what next to do is unclear..
What / where do I try both Models to login a user/client ?

thanks for your help!

Yes using the same table for both causes undesirable results.
I should have mentioned that I'm using Filament and Shield
